vhost-user-backend/src/event_loop.rs

```diff
@@ -3,52 +3,98 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
+use std::collections::HashSet;
 use std::fmt::{Display, Formatter};
 use std::io::{self, Result};
 use std::marker::PhantomData;
 use std::os::unix::io::{AsRawFd, RawFd};
+use std::sync::Mutex;
 
-use vmm_sys_util::epoll::{ControlOperation, Epoll, EpollEvent, EventSet};
+use mio::event::Event;
+use mio::unix::SourceFd;
+use mio::{Events, Interest, Poll, Registry, Token};
 use vmm_sys_util::eventfd::EventFd;
 
 use super::backend::VhostUserBackend;
 use super::vring::VringT;
+use bitflags::bitflags;
 
 /// Errors related to vring epoll event handling.
 #[derive(Debug)]
-pub enum VringEpollError {
+pub enum VringPollError {
     /// Failed to create epoll file descriptor.
-    EpollCreateFd(io::Error),
+    PollerCreate(io::Error),
     /// Failed while waiting for events.
-    EpollWait(io::Error),
+    PollerWait(io::Error),
     /// Could not register exit event
     RegisterExitEvent(io::Error),
     /// Failed to read the event from kick EventFd.
     HandleEventReadKick(io::Error),
     /// Failed to handle the event from the backend.
     HandleEventBackendHandling(io::Error),
+    /// Failed to clone registry.
+    RegistryClone(io::Error),
 }
 
-impl Display for VringEpollError {
+impl Display for VringPollError {
     fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
         match self {
-            VringEpollError::EpollCreateFd(e) => write!(f, "cannot create epoll fd: {e}"),
-            VringEpollError::EpollWait(e) => write!(f, "failed to wait for epoll event: {e}"),
-            VringEpollError::RegisterExitEvent(e) => write!(f, "cannot register exit event: {e}"),
-            VringEpollError::HandleEventReadKick(e) => {
+            VringPollError::PollerCreate(e) => write!(f, "cannot create poller: {e}"),
+            VringPollError::PollerWait(e) => write!(f, "failed to wait for poller event: {e}"),
+            VringPollError::RegisterExitEvent(e) => write!(f, "cannot register exit event: {e}"),
+            VringPollError::HandleEventReadKick(e) => {
                 write!(f, "cannot read vring kick event: {e}")
             }
-            VringEpollError::HandleEventBackendHandling(e) => {
-                write!(f, "failed to handle epoll event: {e}")
+            VringPollError::HandleEventBackendHandling(e) => {
+                write!(f, "failed to handle poll event: {e}")
             }
+            VringPollError::RegistryClone(e) => write!(f, "cannot clone poller's registry: {e}"),
         }
     }
 }
 
-impl std::error::Error for VringEpollError {}
+impl std::error::Error for VringPollError {}
 
 /// Result of vring epoll operations.
-pub type VringEpollResult<T> = std::result::Result<T, VringEpollError>;
+pub type VringEpollResult<T> = std::result::Result<T, VringPollError>;
+
+bitflags! {
+    #[derive(Debug, PartialEq, PartialOrd)]
+    pub struct EventSet: u32 {
+        const READABLE = 1u32;
+        const WRITABLE = 2u32;
+    }
+}
```

Comment:
the `bitflag!` has a "mismatch" with `mio::Interest` you cannot create an empty `mio::Interest` but you can do `EventSet::empty()` (that's the reason you need `.expect()` later). So I think it's better/simpler to just use an enum, like:

```rust
pub enum EventSet {
    Readable,
    Writable,
    All,
}
```

after this, I'm not sure if `EventSet` is the right name